### ISSUE TITLE
Using 2.x and 3.x in parallel

### ISSUE THREAD

**Issue by aik099** [auraphp/Aura.Sql#187]
The project, where Aura.Sql is being used must work on different PHP versions, and on some only Aura.Sql 2.x is supported. 

I have several issues caused by this:

1. the `ProfilerInterface` was moved to a different namespace (I've mitigated that using the `class_alias` function)
2. new methods added on `ProfilerInterface`, that I have no idea how to implement
3. the `Aura\Sql\Iterator\AllIterator` class was removed
4. the `Aura\Sql\Iterator\ColIterator` class was removed

**Comment by brandonsavage** [comment]
Is your intent here to have a version of 2.x installed alongside 3.x and picking which one to use based on which version of PHP you're using?

**Comment by aik099** [comment]
The goal is to have a single code base for my project, that will auto-pick (via Composer) compatible Aura.Sql version based on the used PHP version and my project would work with either of picked Aura.Sql versions.

**Comment by brandonsavage** [comment]
Have you tried the OR (||) operator with two package versions? Composer
should pick between them.

**Comment by djmattyg007** [comment]
Composer should definitely do that automatically for you.

**Comment by aik099** [comment]
Composer is picking the correct Aura.Sql package version, however, due to BC breaks in Aura.Sql 3.x my code, that worked with Aura.Sql 2.x, no longer works on Aura.Sql 3.x.

I even can't adapt my code to work with Aura.Sql 3.x due to the above (in issue description) mentioned BC breaks, that I have no idea how to overcome:

* new `ProfilerInterface` methods with no implementation examples;
* removed iterator classes, which I've found very useful.

**Comment by djmattyg007** [comment]
You're probably going to have to polyfill those things yourself. It's unreasonable to expect an author to maintain backwards compatibility at all costs. That's why they've bumped the major version number of the package.

**Comment by aik099** [comment]
I understand, that I must polyfill them, but I have no idea how. Knowing why things were removed and how it's done in the Aura.Sql 3.x version would be really helpful.

**Comment by brandonsavage** [comment]
Hey Alex,

I'd define an interface and then have concrete classes that proxy to the 2.x and 3.x series components, or replace them. Make sense?

If you're still having trouble, your question is probably out of scope for a GitHub issue, but would fall into the "consulting/contracting" space. I'm sure @pmjones or someone like him could offer some help in this area. Most of us have our email addresses in our profiles.

**Comment by aik099** [comment]
Thanks for explaining. Architectural problems aside I still have no idea how to adapt my code to work with Aura.Sql 3.x with `ProfilerInterface` change and iterator removal.

// cc @pmjones

**Comment by brandonsavage** [comment]
@aik099 It looks like you may have to backport `ProfilerInterface` yourself, or copy the component from the older version. At this point I'm going to close the issue because there's not too much we can specifically do to assist, but I hope you're able to get what you need working.
